File: 560 subarray sum equals k/SubarraySumEqualsK.java
public class SubarraySumEqualsK {
    public int subarraySum(int[] nums, int k) {
        Map<Integer, Integer> map = new HashMap<>();
        int curSum = 0;
        int res = 0;
        for (int num : nums) {
            map.put(curSum, map.getOrDefault(curSum, 0) + 1);
            curSum += num;
            if (map.containsKey(curSum - k)) {
                res += map.get(curSum - k);
            }
        }
        return res;
    }
}

File: 560 subarray sum equals k/SubarraySumEqualsK.py
class SubarraySumEqualsK(object):
    def subarraySum1(self, nums, k):
        """
        :type nums: List[int]
        :type k: int
        :rtype: int
        """
        pre_sums = {}
        cur_sum = 0
        res = 0
        for num in nums:
            if cur_sum in pre_sums:
                pre_sums[cur_sum] += 1
            else:
                pre_sums[cur_sum] = 1
            
            cur_sum += num
            
            if cur_sum - k in pre_sums:
                res += pre_sums[cur_sum - k]
        
        return res

    def subarraySum2(self, nums, k):
        """
        :type nums: List[int]
        :type k: int
        :rtype: int
        """
        pre_sums = {}
        cur_sum = 0
        res = 0
        for num in nums:
            pre_sums[cur_sum] = pre_sums.get(cur_sum, 0) + 1
            cur_sum += num
            res += pre_sums.get(cur_sum - k, 0)
        
        return res